designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+- Repo: eslint/eslint
+- Start Date: (2022-02-28)
+- RFC PR: https://github.com/eslint/rfcs/pull/88
+- Authors: aladdin-add(weiran.zsd@outlook.com)
+
+# convert eslint codebase to esm
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+Convert the ESLint codebase to ESM and publish the ESM version.
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+The Node.js community has been written commonjs for a long time, the same goes for ESLint. But with ESM introduced in ES6 and officially supported by Node.js v12, more and more projects are moving to ESM. Not long ago, we had converted a few repos under the ESLint organization to ESM:
+* @eslint/eslintrc
+* espree
+* eslint-scope
+* eslint-visitor-keys
+* @eslint/create-config
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+There are two main ways to achieve this:
+1. esm-only: write esm, publish esm
+```js
+{
+    "type": "module",
+    "exports": {
+        "./package.json": "./package.json",
+        ".": "./lib/api.js",
+        "./use-at-your-own-risk": "./lib/unsupported-api.js"
+    }
+}
+```
+2. dual-mode: compile esm to cjs, publish esm+cjs
+```js
+    "type": "module",
+    "exports": {
+        "./package.json": "./package.json",
+        ".": {
+            "import": "./lib/api.js",
+            "require": "./dist/api.cjs"
+        },
+        "./use-at-your-own-risk": {
+            "import": "./lib/unsupported-api.js",
+            "require": "./dist/unsupported-api.cjs",
+        }
+    }
+```
+
+Node.js ESM has a limitation: ESM cannot be required. Approach 2 provides a good compatibility for existing projects that rely on ESLint without requiring users to adopt ESM as well. In previous repos, dual-mode was mostly used (except for '@eslint/create-config').
+
+However, dual-mode also has a few disadvantages:
+1. dependencies. Some ESLint's dependencies have been (or will be) esm-only, which means we won't be able to upgrade them if we continue to support CJS. For new dependencies, we cannot use them if they are esm-only.
```

Comment:
A dep going ESM-only is user-hostile - not nearly as bad as one that violates semver, but in a similar bucket. If eslint is depending on packages that have done this, and there truly aren't alternatives in the ecosystem, I'll be happy to create and maintain them for as long as eslint needs them.

Reply:
It would be good if all the dependents go to esm, but it will take a very very long time. 😄

Reply:
That will be the case whether eslint goes ESM-only or not - it can’t be forcibly accelerated.